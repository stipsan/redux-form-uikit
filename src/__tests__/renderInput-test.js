import renderer from 'react-test-renderer'

import RenderInput from '../renderInput'

// Capitalize to render it outside of redux-form

describe('renderInput', () => {
  // @TODO investigate if we can just grab the default props directly from redux-form/Field
  const defaultProps = {
    input: {
      name: '',
      onBlur: () => {},
      onChange: () => {},
      onDragStart: () => {},
      onDrop: () => {},
      onFocus: () => {},
      value: '',
    },
    meta: {
      asyncValidating: false,
      dirty: false,
      invalid: true,
      pristine: true,
      submitting: false,
      touched: false,
      valid: false,
    },
  }
  const props = {
    ...defaultProps,
    input: {
      ...defaultProps.input,
      name: 'email',
    },
    placeholder: 'E-mail',
  }
  it('renders correctly', () => {
    expect(renderer.create(
      <RenderInput {...props} />
    ).toJSON()).toMatchSnapshot()

    expect(renderer.create(
      <RenderInput
        {...props}
        inline
        label="E-mail"
        type="email"
      />
    ).toJSON()).toMatchSnapshot()
  })

  it('renders error messages correctly', () => {
    expect(renderer.create(
      <RenderInput
        {...props}
        meta={{
          ...defaultProps.meta,
          error: 'Shows no error if untouched.',
        }}
      />
    ).toJSON()).toMatchSnapshot()

    expect(renderer.create(
      <RenderInput
        {...props}
        meta={{
          ...defaultProps.meta,
          touched: true,
          error: 'Shows error if touched.',
        }}
      />
    ).toJSON()).toMatchSnapshot()

    expect(renderer.create(
      <RenderInput
        {...props}
        errorClassName="uk-text-danger"
        meta={{
          ...defaultProps.meta,
          touched: true,
          error: 'Allows custom className on error container.',
        }}
      />
    ).toJSON()).toMatchSnapshot()

    expect(renderer.create(
      <RenderInput
        {...props}
        errorDisplay="block"
        meta={{
          ...defaultProps.meta,
          touched: true,
          error: 'Support rendering error message in a block beneath the field.',
        }}
      />
    ).toJSON()).toMatchSnapshot()

    expect(renderer.create(
      <RenderInput
        {...props}
        label="E-mail"
        meta={{
          ...defaultProps.meta,
          touched: true,
          error: 'Neatly renders error in input control group.',
        }}
      />
    ).toJSON()).toMatchSnapshot()

    expect(renderer.create(
      <RenderInput
        {...props}
        errorDisplay="block"
        label="E-mail"
        meta={{
          ...defaultProps.meta,
          touched: true,
          error: 'Support rendering error message in a block beneath the field.',
        }}
      />
    ).toJSON()).toMatchSnapshot()
  })

  it('renders help messages correctly', () => {
    expect(renderer.create(
      <RenderInput
        {...props}
        help="Friendly message to delight the human."
      />
    ).toJSON()).toMatchSnapshot()

    expect(renderer.create(
      <RenderInput
        {...props}
        help="Supports custom className on help message."
        helpClassName="uk-text-warning"
      />
    ).toJSON()).toMatchSnapshot()

    expect(renderer.create(
      <RenderInput
        {...props}
        help="Possible to render below field."
        helpDisplay="block"
      />
    ).toJSON()).toMatchSnapshot()

    expect(renderer.create(
      <RenderInput
        {...props}
        help={<strong>Help attribute can render anything <code>React</code> can render.</strong>}
      />
    ).toJSON()).toMatchSnapshot()
  })

  it('prioritizes error over help', () => {
    expect(renderer.create(
      <RenderInput
        {...props}
        help="Don't render this!"
        meta={{
          ...defaultProps.meta,
          touched: true,
          error: 'Errors have precedence.',
        }}
      />
    ).toJSON()).toMatchSnapshot()

    expect(renderer.create(
      <RenderInput
        {...props}
        errorDisplay="block"
        help="Don't render this!"
        helpDisplay="block"
        meta={{
          ...defaultProps.meta,
          touched: true,
          error: 'Errors have precedence.',
        }}
      />
    ).toJSON()).toMatchSnapshot()

    expect(renderer.create(
      <RenderInput
        {...props}
        help="Don't render this!"
        label="E-mail"
        meta={{
          ...defaultProps.meta,
          touched: true,
          error: 'Errors have precedence.',
        }}
      />
    ).toJSON()).toMatchSnapshot()

    expect(renderer.create(
      <RenderInput
        {...props}
        errorDisplay="block"
        help="Don't render this!"
        helpDisplay="block"
        label="E-mail"
        meta={{
          ...defaultProps.meta,
          touched: true,
          error: 'Errors have precedence.',
        }}
      />
    ).toJSON()).toMatchSnapshot()
  })

  it('supports custom wrapper className', () => {
    expect(renderer.create(
      <RenderInput
        {...props}
        wrapperClassName="uk-display-inline-block"
      />
    ).toJSON()).toMatchSnapshot()
  })

  it('allows overriding defaults', () => {
    expect(renderer.create(
      <RenderInput
        {...props}
        autoComplete={false}
        id="email-confirm"
        label="E-mail"
        placeholder="Confirm E-mail"
      />
    ).toJSON()).toMatchSnapshot()
  })

  it('prevents overriding the danger prop on <Input />', () => {
    expect(renderer.create(
      <RenderInput
        {...props}
        danger
      />
    ).toJSON()).toMatchSnapshot()

    expect(renderer.create(
      <RenderInput
        {...props}
        danger={false}
        meta={{
          ...defaultProps.meta,
          touched: true,
          error: 'Required',
        }}
      />
    ).toJSON()).toMatchSnapshot()
  })

  it('renders asyncValidating status correctly', () => {
    expect(renderer.create(
      <RenderInput
        {...props}
        meta={{
          ...defaultProps.meta,
          asyncValidating: true,
          touched: true,
          pristine: false,
        }}
      />
    ).toJSON()).toMatchSnapshot()

    expect(renderer.create(
      <RenderInput
        {...props}
        flip
        icon="envelope"
        meta={{
          ...defaultProps.meta,
          asyncValidating: true,
          touched: true,
          pristine: false,
        }}
      />
    ).toJSON()).toMatchSnapshot()
  })
})
