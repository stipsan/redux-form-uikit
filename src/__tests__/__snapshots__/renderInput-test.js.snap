exports[`renderInput allows overriding defaults 1`] = `
<div
  className="uk-form-row">
  <label
    className="uk-form-label"
    htmlFor="email-confirm">
    E-mail
  </label>
  <div
    className="uk-form-controls">
    <input
      autoComplete={false}
      className=""
      id="email-confirm"
      name="email"
      onBlur={[Function]}
      onChange={[Function]}
      onDragStart={[Function]}
      onDrop={[Function]}
      onFocus={[Function]}
      placeholder="Confirm E-mail"
      value="" />
  </div>
</div>
`;

exports[`renderInput prevents overriding the danger prop on <Input /> 1`] = `
<div
  className="uk-form-row">
  <input
    autoComplete="email"
    className=""
    id={undefined}
    name="email"
    onBlur={[Function]}
    onChange={[Function]}
    onDragStart={[Function]}
    onDrop={[Function]}
    onFocus={[Function]}
    placeholder="E-mail"
    value="" />
</div>
`;

exports[`renderInput prevents overriding the danger prop on <Input /> 2`] = `
<div
  className="uk-form-row">
  <input
    autoComplete="email"
    className="uk-form-danger"
    id={undefined}
    name="email"
    onBlur={[Function]}
    onChange={[Function]}
    onDragStart={[Function]}
    onDrop={[Function]}
    onFocus={[Function]}
    placeholder="E-mail"
    value="" />
  <p
    className="uk-form-help-inline">
    Required
  </p>
</div>
`;

exports[`renderInput prioritizes error over help 1`] = `
<div
  className="uk-form-row">
  <input
    autoComplete="email"
    className="uk-form-danger"
    id={undefined}
    name="email"
    onBlur={[Function]}
    onChange={[Function]}
    onDragStart={[Function]}
    onDrop={[Function]}
    onFocus={[Function]}
    placeholder="E-mail"
    value="" />
  <p
    className="uk-form-help-inline">
    Errors have precedence.
  </p>
</div>
`;

exports[`renderInput prioritizes error over help 2`] = `
<div
  className="uk-form-row">
  <input
    autoComplete="email"
    className="uk-form-danger"
    id={undefined}
    name="email"
    onBlur={[Function]}
    onChange={[Function]}
    onDragStart={[Function]}
    onDrop={[Function]}
    onFocus={[Function]}
    placeholder="E-mail"
    value="" />
  <p
    className="uk-form-help-block">
    Errors have precedence.
  </p>
</div>
`;

exports[`renderInput prioritizes error over help 3`] = `
<div
  className="uk-form-row">
  <label
    className="uk-form-label"
    htmlFor="email">
    E-mail
  </label>
  <div
    className="uk-form-controls">
    <input
      autoComplete="email"
      className="uk-form-danger"
      id="email"
      name="email"
      onBlur={[Function]}
      onChange={[Function]}
      onDragStart={[Function]}
      onDrop={[Function]}
      onFocus={[Function]}
      placeholder="E-mail"
      value="" />
    <p
      className="uk-form-help-inline">
      Errors have precedence.
    </p>
  </div>
</div>
`;

exports[`renderInput prioritizes error over help 4`] = `
<div
  className="uk-form-row">
  <label
    className="uk-form-label"
    htmlFor="email">
    E-mail
  </label>
  <div
    className="uk-form-controls">
    <input
      autoComplete="email"
      className="uk-form-danger"
      id="email"
      name="email"
      onBlur={[Function]}
      onChange={[Function]}
      onDragStart={[Function]}
      onDrop={[Function]}
      onFocus={[Function]}
      placeholder="E-mail"
      value="" />
    <p
      className="uk-form-help-block">
      Errors have precedence.
    </p>
  </div>
</div>
`;

exports[`renderInput renders asyncValidating status correctly 1`] = `
<div
  className="uk-form-row">
  <div
    className="uk-form-icon">
    <i
      className="uk-icon-spinner uk-icon-spin" />
    <input
      autoComplete="email"
      className=""
      id={undefined}
      name="email"
      onBlur={[Function]}
      onChange={[Function]}
      onDragStart={[Function]}
      onDrop={[Function]}
      onFocus={[Function]}
      placeholder="E-mail"
      value="" />
  </div>
</div>
`;

exports[`renderInput renders asyncValidating status correctly 2`] = `
<div
  className="uk-form-row">
  <div
    className="uk-form-icon uk-form-icon-flip">
    <i
      className="uk-icon-spinner uk-icon-spin" />
    <input
      autoComplete="email"
      className=""
      id={undefined}
      name="email"
      onBlur={[Function]}
      onChange={[Function]}
      onDragStart={[Function]}
      onDrop={[Function]}
      onFocus={[Function]}
      placeholder="E-mail"
      value="" />
  </div>
</div>
`;

exports[`renderInput renders correctly 1`] = `
<div
  className="uk-form-row">
  <input
    autoComplete="email"
    className=""
    id={undefined}
    name="email"
    onBlur={[Function]}
    onChange={[Function]}
    onDragStart={[Function]}
    onDrop={[Function]}
    onFocus={[Function]}
    placeholder="E-mail"
    value="" />
</div>
`;

exports[`renderInput renders correctly 2`] = `
<div
  className="uk-display-inline-block">
  <label
    className="uk-form-label"
    htmlFor="email">
    E-mail
  </label>
  <div
    className="uk-form-controls">
    <input
      autoComplete="email"
      className=""
      id="email"
      name="email"
      onBlur={[Function]}
      onChange={[Function]}
      onDragStart={[Function]}
      onDrop={[Function]}
      onFocus={[Function]}
      placeholder="E-mail"
      type="email"
      value="" />
  </div>
</div>
`;

exports[`renderInput renders error messages correctly 1`] = `
<div
  className="uk-form-row">
  <input
    autoComplete="email"
    className=""
    id={undefined}
    name="email"
    onBlur={[Function]}
    onChange={[Function]}
    onDragStart={[Function]}
    onDrop={[Function]}
    onFocus={[Function]}
    placeholder="E-mail"
    value="" />
</div>
`;

exports[`renderInput renders error messages correctly 2`] = `
<div
  className="uk-form-row">
  <input
    autoComplete="email"
    className="uk-form-danger"
    id={undefined}
    name="email"
    onBlur={[Function]}
    onChange={[Function]}
    onDragStart={[Function]}
    onDrop={[Function]}
    onFocus={[Function]}
    placeholder="E-mail"
    value="" />
  <p
    className="uk-form-help-inline">
    Shows error if touched.
  </p>
</div>
`;

exports[`renderInput renders error messages correctly 3`] = `
<div
  className="uk-form-row">
  <input
    autoComplete="email"
    className="uk-form-danger"
    id={undefined}
    name="email"
    onBlur={[Function]}
    onChange={[Function]}
    onDragStart={[Function]}
    onDrop={[Function]}
    onFocus={[Function]}
    placeholder="E-mail"
    value="" />
  <p
    className="uk-form-help-inline uk-text-danger">
    Allows custom className on error container.
  </p>
</div>
`;

exports[`renderInput renders error messages correctly 4`] = `
<div
  className="uk-form-row">
  <input
    autoComplete="email"
    className="uk-form-danger"
    id={undefined}
    name="email"
    onBlur={[Function]}
    onChange={[Function]}
    onDragStart={[Function]}
    onDrop={[Function]}
    onFocus={[Function]}
    placeholder="E-mail"
    value="" />
  <p
    className="uk-form-help-block">
    Support rendering error message in a block beneath the field.
  </p>
</div>
`;

exports[`renderInput renders error messages correctly 5`] = `
<div
  className="uk-form-row">
  <label
    className="uk-form-label"
    htmlFor="email">
    E-mail
  </label>
  <div
    className="uk-form-controls">
    <input
      autoComplete="email"
      className="uk-form-danger"
      id="email"
      name="email"
      onBlur={[Function]}
      onChange={[Function]}
      onDragStart={[Function]}
      onDrop={[Function]}
      onFocus={[Function]}
      placeholder="E-mail"
      value="" />
    <p
      className="uk-form-help-inline">
      Neatly renders error in input control group.
    </p>
  </div>
</div>
`;

exports[`renderInput renders error messages correctly 6`] = `
<div
  className="uk-form-row">
  <label
    className="uk-form-label"
    htmlFor="email">
    E-mail
  </label>
  <div
    className="uk-form-controls">
    <input
      autoComplete="email"
      className="uk-form-danger"
      id="email"
      name="email"
      onBlur={[Function]}
      onChange={[Function]}
      onDragStart={[Function]}
      onDrop={[Function]}
      onFocus={[Function]}
      placeholder="E-mail"
      value="" />
    <p
      className="uk-form-help-block">
      Support rendering error message in a block beneath the field.
    </p>
  </div>
</div>
`;

exports[`renderInput renders help messages correctly 1`] = `
<div
  className="uk-form-row">
  <input
    autoComplete="email"
    className=""
    id={undefined}
    name="email"
    onBlur={[Function]}
    onChange={[Function]}
    onDragStart={[Function]}
    onDrop={[Function]}
    onFocus={[Function]}
    placeholder="E-mail"
    value="" />
  <p
    className="uk-form-help-inline">
    Friendly message to delight the human.
  </p>
</div>
`;

exports[`renderInput renders help messages correctly 2`] = `
<div
  className="uk-form-row">
  <input
    autoComplete="email"
    className=""
    id={undefined}
    name="email"
    onBlur={[Function]}
    onChange={[Function]}
    onDragStart={[Function]}
    onDrop={[Function]}
    onFocus={[Function]}
    placeholder="E-mail"
    value="" />
  <p
    className="uk-form-help-inline uk-text-warning">
    Supports custom className on help message.
  </p>
</div>
`;

exports[`renderInput renders help messages correctly 3`] = `
<div
  className="uk-form-row">
  <input
    autoComplete="email"
    className=""
    id={undefined}
    name="email"
    onBlur={[Function]}
    onChange={[Function]}
    onDragStart={[Function]}
    onDrop={[Function]}
    onFocus={[Function]}
    placeholder="E-mail"
    value="" />
  <p
    className="uk-form-help-block">
    Possible to render below field.
  </p>
</div>
`;

exports[`renderInput renders help messages correctly 4`] = `
<div
  className="uk-form-row">
  <input
    autoComplete="email"
    className=""
    id={undefined}
    name="email"
    onBlur={[Function]}
    onChange={[Function]}
    onDragStart={[Function]}
    onDrop={[Function]}
    onFocus={[Function]}
    placeholder="E-mail"
    value="" />
  <p
    className="uk-form-help-inline">
    <strong>
      Help attribute can render anything 
      <code>
        React
      </code>
       can render.
    </strong>
  </p>
</div>
`;

exports[`renderInput supports custom wrapper className 1`] = `
<div
  className="uk-display-inline-block uk-form-row">
  <input
    autoComplete="email"
    className=""
    id={undefined}
    name="email"
    onBlur={[Function]}
    onChange={[Function]}
    onDragStart={[Function]}
    onDrop={[Function]}
    onFocus={[Function]}
    placeholder="E-mail"
    value="" />
</div>
`;
